Source/Session.swift

```diff
@@ -922,26 +922,21 @@ open class Session {
 
     func updateStatesForTask(_ task: URLSessionTask, request: Request) {
         request.withState { state in
-            switch (startRequestsImmediately, state) {
-            case (true, .initialized):
-                rootQueue.async { request.resume() }
-            case (false, .initialized):
+            switch state {
+            case .initialized, .finished:
                 // Do nothing.
                 break
-            case (_, .resumed):
+            case .resumed:
                 task.resume()
                 rootQueue.async { request.didResumeTask(task) }
-            case (_, .suspended):
+            case .suspended:
                 task.suspend()
                 rootQueue.async { request.didSuspendTask(task) }
-            case (_, .cancelled):
+            case .cancelled:
                 // Resume to ensure metrics are gathered.
                 task.resume()
                 task.cancel()
                 rootQueue.async { request.didCancelTask(task) }
-            case (_, .finished):
-                // Do nothing
-                break
```

Comment:
Great updates here. Thankfully the huge amount of tests we have for all these scenarios can confirm these changes are valid. In theory though, you've eliminated the entire race by simply waiting to resume the request until the response serializer has been added. This also fixes the race condition for hitting the URLCache faster than the response serializer is added.